### PR TITLE
Attester Double Vote Detection for Optimized Slasher

### PR DESCRIPTION
**Pull request by rauljordan**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

This PR adds double attester vote detection for our optimized slasher, which is fairly straightforward.

**Which issues(s) does this PR fix?**

Part of #8331

### CODE REVIEW

File: beacon-chain/db/kv/slasher.go [hunk 1]
```diff
@@ -70,6 +70,46 @@ func (s *Store) SaveLatestEpochAttestedForValidators(
 	})
 }
 
+// CheckDoubleAttesterVotes retries any slashable double votes that exist
+// for a series of input attestations.
+func (s *Store) CheckAttesterDoubleVotes(
+	ctx context.Context, attestations []*slashertypes.CompactAttestation,
+) ([]*slashertypes.AttesterDoubleVote, error) {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.CheckAndUpdateAttestationRecordForValidators")
```
Comment (terencechain) [inline beacon-chain/db/kv/slasher.go:78]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "BeaconDB.CheckAttesterDoubleVotes")
```

File: beacon-chain/slasher/detect_attestations.go [hunk 2]
```diff
@@ -71,6 +81,69 @@ func (s *Service) detectSlashableAttestations(
...
 }
 
+// Check for attester slashing double votes by looking at every single validator index
+// in each attestation's attesting indices and checking if there already exist records for such
+// attestation's target epoch. If so, we append a double vote slashing object to a list of slashings
+// we return to the caller.
+func (s *Service) checkDoubleVotes(
+	ctx context.Context, attestations []*slashertypes.CompactAttestation,
+) ([]*slashertypes.Slashing, error) {
+	// We check if there are any slashable double votes in the input list
```
Comment (terencechain) [inline beacon-chain/slasher/detect_attestations.go:91]: Good candidate for span tracing?

File: beacon-chain/slasher/detect_attestations.go [hunk 3]
```diff
@@ -71,6 +81,69 @@ func (s *Service) detectSlashableAttestations(
...
+		return nil, err
+	}
+	return append(slashings, moreSlashings...), nil
+}
+
+// Check for double votes in our database given a list of incoming attestations.
+func (s *Service) checkDoubleVotesOnDisk(
+	ctx context.Context, attestations []*slashertypes.CompactAttestation,
+) ([]*slashertypes.Slashing, error) {
+	doubleVotes, err := s.serviceCfg.Database.CheckAttesterDoubleVotes(
```
Comment (terencechain) [inline beacon-chain/slasher/detect_attestations.go:128]: Good candidate for span tracing?

File: beacon-chain/slasher/detect_attestations.go [hunk 4]
```diff
@@ -24,31 +24,38 @@ type chunkUpdateArgs struct {
...
-// 3. Update the latest written epoch for all validators involved to the current epoch.
+// 4. Update the latest written epoch for all validators involved to the current epoch.
 //
 // This function performs a lot of critical actions and is split into smaller helpers for cleanliness.
 func (s *Service) detectSlashableAttestations(
 	ctx context.Context,
 	args *chunkUpdateArgs,
 	attestations []*slashertypes.CompactAttestation,
 ) error {
+	slashings := make([]*slashertypes.Slashing, 0)
```
Comment (terencechain) [inline beacon-chain/slasher/detect_attestations.go:39]: You can move this line down to `// Consolidate all slashings into a slice...`

File: beacon-chain/slasher/detect_attestations.go [hunk 5]
```diff
@@ -57,11 +64,14 @@ func (s *Service) detectSlashableAttestations(
...
 
-	totalSlashings := append(slashings, moreSlashings...)
-	if len(totalSlashings) > 0 {
-		log.WithField("numSlashings", len(totalSlashings)).Info("Slashable offenses found")
+	// Consolidate all slashings into a slice.
+	slashings = append(slashings, doubleVoteSlashings...)
+	slashings = append(slashings, surroundingSlashings...)
+	slashings = append(slashings, surroundedSlashings...)
+	if len(slashings) > 0 {
+		log.WithField("numSlashings", len(slashings)).Info("Slashable offenses found")
```
Comment (terencechain) [inline beacon-chain/slasher/detect_attestations.go:72]: I think we can log more info here. What kind of offense and which validator index committed the offense?

File: beacon-chain/slasher/detect_attestations.go [hunk 6]
```diff
@@ -71,6 +81,69 @@ func (s *Service) detectSlashableAttestations(
 	return s.serviceCfg.Database.SaveLatestEpochAttestedForValidators(ctx, validatorIndices, args.currentEpoch)
 }
 
+// Check for attester slashing double votes by looking at every single validator index
+// in each attestation's attesting indices and checking if there already exist records for such
+// attestation's target epoch. If so, we append a double vote slashing object to a list of slashings
+// we return to the caller.
+func (s *Service) checkDoubleVotes(
```
Comment (terencechain) [inline beacon-chain/slasher/detect_attestations.go:88]: Im afraid `detect_attestations.go ` will get too big, I'd consider breaking them down into smaller files:
`detech_double_votes.go`, `detech_surround_votes.go`... etc

File: beacon-chain/slasher/detect_attestations.go [hunk 7]
```diff
@@ -71,6 +81,69 @@ func (s *Service) detectSlashableAttestations(
...
+	ctx context.Context, attestations []*slashertypes.CompactAttestation,
+) ([]*slashertypes.Slashing, error) {
+	// We check if there are any slashable double votes in the input list
+	// of attestations with respect to each other.
+	slashings := make([]*slashertypes.Slashing, 0)
+	existingAtts := make(map[string][32]byte)
+	for _, att := range attestations {
+		for _, valIdx := range att.AttestingIndices {
+			key := fmt.Sprintf("%d:%d", att.Target, valIdx)
```
Comment (terencechain) [inline beacon-chain/slasher/detect_attestations.go:97]: `Sprintf` is the slowest among different ways concatenate. Given the large number of attesting indices per slot, we should be frugal here.

https://dev.to/pmalhaire/concatenate-strings-in-golang-a-quick-benchmark-4ahh

File: beacon-chain/slasher/detect_attestations.go (hunk 5 above)
Comment (terencechain) [reply beacon-chain/slasher/detect_attestations.go:72]: Nvm, it looks like we do log the more details after

File: beacon-chain/slasher/detect_attestations.go (hunk 6 above)
Comment (rauljordan) [reply beacon-chain/slasher/detect_attestations.go:89]: I can do this in a separate PR, im afraid the diff will then be too big in this PR

**Review by terencechain — APPROVED** [review]

**Pull request merged**